docs/proposals/cli.md

```diff
@@ -0,0 +1,184 @@
+reviewers:
+  - "@gabemontero"
+approvers:
+  - TBD
+creation-date: 2020-10-07
+last-updated:  2020-10-08
+status: provisional
+---
+
+CLI
+---
+
+# Summary
+
+This enhancement proposal describes a command-line client for Shipwright Build operator, covering
+from the command line structure, the usage as a kubectl plugin, and as a standalone binary, as
+well as testing this new project.
+
+# Motivation
+
+Command-line interface is a fundamental component for developer experience and modern automation.
+This enhancement proposal describes a command-line interface, i.e. "CLI", to enhance the developer
+experience and provide easier means to interact and automate Shipwright's Builds.
+
+## Goals
+
+*   Improve the user experience by providing a specialized client
+*   Make possible extended automation with shell scripting
+
+## Non-Goals
+
+*   Re-implement operator logic
+*   Become the mandatory API client
+
+# Proposal
+
+Create a command-line client for Shipwright Build to improve the user experience and also improve
+the ability to automate workflows.
+
+This new project will concentrate our effects on the API client side of the Shipwright project,
+and by maintaining the expected separation of `cmd` and `pkg`, developers can embed and extend the
+possibilities of consuming Shipwright Operator API.
+
+A [proof-of-concept command-line interface project](https://github.com/otaviof/shp) has been
+written to scaffold the features, and evaluate the changes proposed in this document. The project
+documentation (`README.md`) is focused on developers and contributors, while the features it will
+provide are contained in this document.
+
+## Usage and Installation
+
+This command-line application will be consumed as a standalone binary, or as a plugin in `kubectl`.
+This implies a few more extra settings, and making sure the application binary complies with the
+[expected convention](https://krew.sigs.k8s.io/docs/developer-guide/develop/best-practices/).
+
+### Standalone
+
+The standalone binary will be available via the usual means, final users will be able to simply
+install with `go get…`, or pick a pre-compiled binary from GitHub release pages. Consider the
+[release section](#Releasing) for more.
+
+### Plugin
+
+As a [krew plugin](https://krew.sigs.k8s.io/docs/developer-guide/), the expected usage is:
+
+```sh
+$ kubectl shp ...
+```
+
+And the following global parameters are expected, to keep consistency with `kubectl`:
+
+*   `-h/--help`
+*   `-n/--namespace`
+*   `-A/--all-namespaces`
+
+Those three parameters are mandatory to keep compatibility. However, the
+[upstream kubectl package](https://pkg.go.dev/k8s.io/kubectl@v0.17.6/pkg/cmd/util) provides
+[more options](https://github.com/otaviof/shp/blob/55ce2e8d58435b0264e3db0bef5cf439abfeca18/vendor/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go#L317)
+out of the box.
+
+## Command
+
+The project will use [cobra](https://github.com/spf13/cobra) for scaffolding the command-line
```

Comment:
Might as well include Viper also https://github.com/spf13/viper for saving configuration files etc.

Reply:
Indeed. Viper comes handy for this type of command-line together with Cobra. Thanks.